// @nrwl/schematics/src/lib-versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.angularCliVersion = '1.7.1';
exports.angularVersion = '5.2.6';
exports.angularJsVersion = '1.6.6';
exports.ngrxVersion = '5.1.0';
exports.ngrxStoreFreezeVersion = '^0.2.1';
exports.routerStoreVersion = '5.0.1';
exports.nxVersion = '0.9.1';
exports.schematicsVersion = '0.9.1';
exports.angularCliSchema = './node_modules/@nrwl/schematics/src/schema.json';
exports.latestMigration = '20180313-add-tags';
exports.prettierVersion = '1.10.2';
exports.typescriptVersion = '2.6.2';
exports.rxjsVersion = '^5.5.6';
exports.devKitCoreVersion = '^0.0.29';
exports.devKitSchematicsVersion = '0.0.52';
exports.schematicsAngularVersion = '0.1.17';
exports.libVersions = {
    angularVersion: exports.angularVersion,
    angularCliVersion: exports.angularCliVersion,
    angularJsVersion: exports.angularJsVersion,
    ngrxVersion: exports.ngrxVersion,
    ngrxStoreFreezeVersion: exports.ngrxStoreFreezeVersion,
    nxVersion: exports.nxVersion,
    schematicsVersion: exports.schematicsVersion,
    prettierVersion: exports.prettierVersion,
    angularCliSchema: exports.angularCliSchema,
    latestMigration: exports.latestMigration,
    typescriptVersion: exports.typescriptVersion,
    rxjsVersion: exports.rxjsVersion,
    devKitCoreVersion: exports.devKitCoreVersion,
    devKitSchematicsVersion: exports.devKitSchematicsVersion,
    schematicsAngularVersion: exports.schematicsAngularVersion,
    routerStoreVersion: exports.routerStoreVersion
};
