// @nrwl/schematics/src/command-line/affected-apps.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var affected_apps_1 = require("./affected-apps");
describe('Calculates Dependencies Between Apps and Libs', function () {
    describe('dependencies', function () {
        it('should return a graph with a key for every project', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: [],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'app2',
                    root: '',
                    files: [],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function () { return null; });
            expect(deps).toEqual({ app1: [], app2: [] });
        });
        it('should infer deps between projects based on imports', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'lib1',
                    root: '',
                    files: ['lib1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                },
                {
                    name: 'lib2',
                    root: '',
                    files: ['lib2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                }
            ], function (file) {
                switch (file) {
                    case 'app1.ts':
                        return "\n            import '@nrwl/lib1';\n            import '@nrwl/lib2/deep';\n          ";
                    case 'lib1.ts':
                        return "import '@nrwl/lib2'";
                    case 'lib2.ts':
                        return '';
                }
            });
            expect(deps).toEqual({
                app1: [
                    { projectName: 'lib1', type: affected_apps_1.DependencyType.es6Import },
                    { projectName: 'lib2', type: affected_apps_1.DependencyType.es6Import }
                ],
                lib1: [{ projectName: 'lib2', type: affected_apps_1.DependencyType.es6Import }],
                lib2: []
            });
        });
        it('should infer dependencies expressed via loadChildren', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'lib1',
                    root: '',
                    files: ['lib1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                },
                {
                    name: 'lib2',
                    root: '',
                    files: ['lib2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                }
            ], function (file) {
                switch (file) {
                    case 'app1.ts':
                        return "\n            const routes = {\n              path: 'a', loadChildren: '@nrwl/lib1#LibModule',\n              path: 'b', loadChildren: '@nrwl/lib2/deep#LibModule'\n            };\n          ";
                    case 'lib1.ts':
                        return '';
                    case 'lib2.ts':
                        return '';
                }
            });
            expect(deps).toEqual({
                app1: [
                    { projectName: 'lib1', type: affected_apps_1.DependencyType.loadChildren },
                    { projectName: 'lib2', type: affected_apps_1.DependencyType.loadChildren }
                ],
                lib1: [],
                lib2: []
            });
        });
        it('should handle non-ts files', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['index.html'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function () { return null; });
            expect(deps).toEqual({ app1: [] });
        });
        it('should handle projects with the names starting with the same string', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'aa',
                    root: '',
                    files: ['aa.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'aa/bb',
                    root: '',
                    files: ['bb.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function (file) {
                switch (file) {
                    case 'aa.ts':
                        return "import '@nrwl/aa/bb'";
                    case 'bb.ts':
                        return '';
                }
            });
            expect(deps).toEqual({
                aa: [{ projectName: 'aa/bb', type: affected_apps_1.DependencyType.es6Import }],
                'aa/bb': []
            });
        });
        it('should not add the same dependency twice', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'aa',
                    root: '',
                    files: ['aa.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'bb',
                    root: '',
                    files: ['bb.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function (file) {
                switch (file) {
                    case 'aa.ts':
                        return "\n              import '@nrwl/bb/bb'\n              import '@nrwl/bb/bb'\n              ";
                    case 'bb.ts':
                        return '';
                }
            });
            expect(deps).toEqual({
                aa: [{ projectName: 'bb', type: affected_apps_1.DependencyType.es6Import }],
                bb: []
            });
        });
        it('should not add a dependency on self', function () {
            var deps = affected_apps_1.dependencies('nrwl', [
                {
                    name: 'aa',
                    root: '',
                    files: ['aa.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function (file) {
                switch (file) {
                    case 'aa.ts':
                        return "\n              import '@nrwl/aa/aa'\n              ";
                }
            });
            expect(deps).toEqual({ aa: [] });
        });
    });
    describe('affectedApps', function () {
        it('should return the list of affected files', function () {
            var affected = affected_apps_1.affectedApps('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'app2',
                    root: '',
                    files: ['app2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'lib1',
                    root: '',
                    files: ['lib1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                },
                {
                    name: 'lib2',
                    root: '',
                    files: ['lib2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                }
            ], function (file) {
                switch (file) {
                    case 'app1.ts':
                        return "\n            import '@nrwl/lib1';\n          ";
                    case 'app2.ts':
                        return "";
                    case 'lib1.ts':
                        return "import '@nrwl/lib2'";
                    case 'lib2.ts':
                        return '';
                }
            }, ['lib2.ts']);
            expect(affected).toEqual(['app1']);
        });
        it('should return app app names if a touched file is not part of a project', function () {
            var affected = affected_apps_1.affectedApps('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'app2',
                    root: '',
                    files: ['app2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'lib1',
                    root: '',
                    files: ['lib1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                }
            ], function (file) {
                switch (file) {
                    case 'app1.ts':
                        return "\n            import '@nrwl/lib1';\n          ";
                    case 'app2.ts':
                        return "";
                    case 'lib1.ts':
                        return "import '@nrwl/lib2'";
                }
            }, ['package.json']);
            expect(affected).toEqual(['app2', 'app1']);
        });
        it('should handle slashes', function () {
            var affected = affected_apps_1.affectedApps('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['one\\app1.ts', 'two/app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function (file) {
                switch (file) {
                    case 'one/app1.ts':
                        return '';
                    case 'two/app1.ts':
                        return '';
                }
            }, ['one/app1.ts', 'two/app1.ts']);
            expect(affected).toEqual(['app1']);
        });
        it('should handle circular dependencies', function () {
            var affected = affected_apps_1.affectedApps('nrwl', [
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'app2',
                    root: '',
                    files: ['app2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                }
            ], function (file) {
                switch (file) {
                    case 'app1.ts':
                        return "import '@nrwl/app2';";
                    case 'app2.ts':
                        return "import '@nrwl/app1';";
                }
            }, ['app1.ts']);
            expect(affected).toEqual(['app2', 'app1']);
        });
    });
    describe('touchedProjects', function () {
        it('should return the list of touchedProjects', function () {
            var tp = affected_apps_1.touchedProjects([
                {
                    name: 'app1',
                    root: '',
                    files: ['app1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'app2',
                    root: '',
                    files: ['app2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.app
                },
                {
                    name: 'lib1',
                    root: '',
                    files: ['lib1.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                },
                {
                    name: 'lib2',
                    root: '',
                    files: ['lib2.ts'],
                    tags: [],
                    type: affected_apps_1.ProjectType.lib
                }
            ], ['lib2.ts', 'app2.ts', 'package.json']);
            expect(tp).toEqual(['lib2', 'app2', null]);
        });
    });
});
